src/geeknn/ordination/_base.py

```diff
@@ -0,0 +1,282 @@
+from numpy.typing import NDArray
+from pydantic import BaseModel
+from sklearn.base import TransformerMixin
+
+from .utils import crosswalk_to_ids, filter_neighbors, return_k_neighbors, scores_to_fc
+
+
+class Geometry(BaseModel):
+    """Client-side proxy for ee.Geometry object."""
+
+    type: str
+    coordinates: list[Any]
+
+
+class Feature(BaseModel):
+    """Client-side proxy for ee.Feature object."""
+
+    type: str
+    geometry: Geometry
+    id: str
+    properties: dict[str, Any]
+
+    def get_property(self, property: str) -> Any:
+        return self.properties.get(property)
+
+    def get_properties(self, properties: list[str]) -> list[Any]:
+        return [self.get_property(x) for x in properties]
+
+
+class FeatureCollection(BaseModel):
+    """Client-side proxy for ee.FeatureCollection object."""
+
+    type: str
+    columns: dict[str, Any]
+    version: int
+    id: str
+    properties: dict[str, Any]
+    features: list[Feature]
+
+    def aggregate_array(self, property: str) -> NDArray:
+        """Return a single property as a 1-D numpy array."""
+        return np.array([f.get_property(property) for f in self.features])
+
+    def properties_to_array(self, properties: list[str]) -> NDArray:
+        """Return specified properties as a 2-D numpy array."""
+        return np.array([f.get_properties(properties) for f in self.features])
+
+    @classmethod
+    def from_ee_feature_collection(
+        cls, fc: ee.FeatureCollection, num_threads: int = -1, chunk_size: int = 5000
+    ):
+        """Create a client-side FeatureCollection from an ee.FeatureCollection
+        using multiple threads and chunking to avoid memory issues.
+        """
+        size = fc.size().getInfo()
+        info = fc.limit(0).getInfo()
+        chunks = [fc.toList(chunk_size, i) for i in range(0, size, chunk_size)]
+
+        with Parallel(n_jobs=num_threads, backend="threading") as p:
+            chunk_data = p(delayed(chunk.getInfo)() for chunk in chunks)
+
+        info.update({"features": list(chain.from_iterable(chunk_data))})
+        return cls(**info)
+
+
+class Raw:
+    def __init__(self, k=1, max_duplicates=None):
+        self.k = k
+        self.max_duplicates = max_duplicates if max_duplicates is not None else 5
+
+    def transform_image(self, X_image: ee.Image) -> ee.Image:
+        """No-op placeholder."""
+        return X_image
+
+    def transform_fc(self, fc: ee.FeatureCollection) -> ee.FeatureCollection:
+        """No-op placeholder."""
+        return fc
+
+    @property
+    def k_nearest(self):
+        """Return the number of requested nearest neighbors plus max_duplicates
+        in order to account for non-independent samples.
+        """
+        return self.k + self.max_duplicates
+
+    def get_ids(self, fc: FeatureCollection, id_field: str) -> list[Any]:
+        """Return the IDs from the feature collection."""
+        return list(map(int, fc.aggregate_array(id_field)))
```

Comment:
Well, they *should* be ints 😄. But your point is well-taken ... Because these IDs are eventually used in an `ee.Image` context (i.e., the neighbor images), things will go wrong if they are not integers. Now that I've deleted the `get_ids` function, I think `utils.scores_to_fc` would actually work (ignoring the type hint of `ids` as `list[int]`), so the error wouldn't actually show up until running predict.

Do you think I need to fail quickly here if the IDs are not integers?

Reply:
Ah, I see! 

> Do you think I need to fail quickly here if the IDs are not integers?

I think it might be worth it, mostly for the benefit of being able to throw a more helpful error than EE probably would. I suppose there's no way to easily validate that field using `pydantic`, since the values for `Feature.properties` need to be `Any` to accommodate other fields, right?

Reply:
> I suppose there's no way to easily validate that field using `pydantic`, since the values for `Feature.properties` need to be `Any` to accommodate other fields, right?

Exactly right.  And I think it's complicated by the fact that I'm getting IDs both client-side (through `FeatureCollection.aggregate_array`) and server-side (through `ee.FeatureCollection.aggregate_array`).  I'd like to use the same logic to validate both, but that will likely require a `.getInfo()`.  I can probably make this a bit faster by assuming the first item is representative of the list and use `.limit(1)` to just retrieve a single record from the server-side collection, but it feels like I'll need to use `isinstance` for type checking the client-side object.

I've added this to the list of issues that I'm not going to address with this PR and will deal with with subsequent (smaller) issues/PRs.